src/service/template-impl.js

```diff
@@ -119,6 +119,33 @@ export class BaseTemplate {
     return singleElement || root;
   }
 
+  /**
+   * Helps the template implementation to unwrap the root element. The root
+   * element contents are unwrapped as an array of elements. Any text node
+   * children are normalized inside a <div>.
+   * @param {!Element} root
+   * @return {!Array<!Element>}
+   * @protected @final
+   */
+  unwrapAsArray(root) {
+    const elements = [];
+    for (let n = root.firstChild; n != null; n = n.nextSibling) {
+      if (n.nodeType == /* TEXT */ 3) {
+        const content = n.textContent.trim();
+        if (content) {
+          const element = this.win.document.createElement('div');
+          element.textContent = content;
+          elements.push(element);
+        }
+      } else if (n.nodeType == /* COMMENT */ 8) {
+        // Ignore comments.
+      } else if (n.nodeType == /* ELEMENT */ 1) {
+        elements.push(dev().assertElement(n));
+      }
+    }
```

Comment:
Can we call `unwrap()` internally instead?

Reply:
Could you elaborate on the idea? Instead of what?

Reply:
Instead of duplicating the code.

Reply:
Are you asking for `forceUnwrap` to use `unwrap` on each of its root's child nodes? Or are you asking for `setHtml` to `unwrap` instead of `forceUnwrap` without any modifications to `unwrap`?

Reply:
To decrease duplicated code, I had `forceUnwrap` call `unwrap` instead of the other way around. LMK if you had something else in mind.

Reply:
Yes, `setHtml` wraps once prior to the purification-wrapping to guarantee the post-purification HTML can be unwrapped to a single element, then it will `unwrapChildren` for the target contents.

So the call order for amp-mustache 0.2 would be:
- `setHtml("abc")`
  - `purifyAndSetHtml_("<div>abc</div>")`
    - `purifyHtml("<div><div>abc</div></div>")`
    - `tryUnwrap(<div><div>abc</div></div>)`
  - `unwrapChildren(<div>abc</div>)`
- finds `abc` as first child, returns it wrapped `<div>abc</div>`

Another example: 
- `setHtml("<a>abc</a>")`
  - `purifyAndSetHtml_("<div><a>abc</a></div>")`
    - `purifyHtml("<div><div><a>abc</a></div></div>")`
    - `tryUnwrap(<div><div><a>abc</a></div></div>)`
  - `unwrapChildren(<div><a>abc</a></div>)`
- finds `<a>abc</a>` as first child, returns `<a>abc</a>`

I see your point in the first example that the `<div>` passed to `unwrapChildren` isn't the same as the `<div>` that gets returned. In that case I added a guard for the child traversal:
```js
if (!root.firstElementChild) {
  return root;
}
```

This can also prevent `unwrapChildren(<a>abc</a>)`  from returning `<div>abc</div>`, which might not be expected at a glance. This edge case isn't particularly applicable to the way it is used now, but it's worth covering in case it is used this way in the future.

Reply:
Thanks for iterating on this given vague direction. IMO still pretty confusing to read... what if we created an iterator helper instead that both functions call?

```js
function unwrap_(root, callback) {
	for (let n = root.firstChild; n != null; n = n.nextSibling) {
	  if (n.nodeType == /* TEXT */ 3) {
	    const text = n.textContent.trim();
	    callback(text);
	  } else if (n.nodeType == /* COMMENT */ 8) {
	    // Ignore comments.
	  } else if (n.nodeType == /* ELEMENT */ 1) {
	    callback(dev().assertElement(n));
	  }
	}
}

tryUnwrap(root) {
	let onlyChild;
	this.unwrap_(root, c => {
		if (onlyChild === undefined && c.nodeType) {
			onlyChild = c;
		} else {
      onlyChild = null;
    }
	});
	return onlyChild;
}

unwrapChildren(root) {
	let children = [];
	this.unwrap_(root, c => {
		if (typeof c == 'string') {
			// Wrap in <div> and push.
		} else {
			children.push(c);
		}
	});
  return children;
}
```

Reply:
Applied your recommendation with the following differences:
- `unwrap_` is renamed to `visitChildren_`
- `callback` is only applied to text nodes when non-empty
- `unwrapChildren` still guards child traversal and returns type `Element` if it only finds one child
- `tryUnwrap` returns `onlyChild || root`

Question: Is there a reason to check `nodeType` in `tryUnwrap` and `typeof` in `unwrapChildren`? Why don't they distinguish their arguments in the same way?

Reply:
Also, in this case are we increasing code size in favor of readabilit